### DISCUSSION TITLE
`hrefTo` function in links addon doesn't need to be async

### DISCUSSION THREAD

**Discussion by JoshTumath** [storybookjs/storybook#23939] [Ideas]
### Is your feature request related to a problem? Please describe.

In my team, we make our own Storybook block components in React to link to other stories, like so. We use the `linkTo` function from the links addon to achieve that.

```jsx
<SignpostLink
  href="#"
  onClick={event => {
    event.preventDefault();
    linkTo(storyNav)();
  }}
>
  {name}
</SignpostLink>
```

We also want to use the `hrefTo` function from the links addon to link to other stories so that users can still middle click or command-click on our links so that they can open the links in a new tab. However, we can't because the `hrefTo` function is async. Rendering asynchronously fetched content in a React component is more hassle, so we don't want to use it.

We looked at the implementation code for the `hrefTo` function, and it doesn't seem as though it's doing anything async anyway. It looks like it could be easily converted to a synchronous function.

https://github.com/storybookjs/storybook/blob/next/code/addons/links/src/utils.ts#L37-L56

### Describe the solution you'd like

I would like the `hrefTo` function to not be async (i.e. not use a Promise).

### Are you able to assist to bring the feature to reality?

no

**Comment by conorriches** [comment]
I've just happened to have stumbled into this problem myself, it would be great to see it resolved

**Comment by Vazaril** [comment]
I also looked into this behaviour of the `hrefTo` function. It's true that the function doesn't use any asynchronous code, but changing it directly via a small fix would certainly break many code dependencies. Maybe a temporary solution in a minor patch would be an introduction of a new `hrefToSync` function for users who want to avoid the async capabilities of the `hrefTo` function. Later major releases could introduce a complete revamp of the function in a more sophisticated manner with underlying tests.

I don't know if feature requests need to be green lit by the Team of maintainers or if anyone can work on an implementation in a separate fork. I could look into this request if an implementation is wanted.

**Reply by JoshTumath** [reply]
Thank you, @Vazaril , that would be great. A new `hrefToSync` function would be perfect. I'd also be concerned about breaking existing implementations by removing the Promise.

**Reply by Vazaril** [reply]
Hey, @JoshTumath  thanks for your reply. I worked on a `hrefToSync` function in my own branch:
- Repo: https://github.com/Vazaril/storybook/tree/feat-links-addon-hreftosync
- Function: https://github.com/Vazaril/storybook/blob/dba7926980a5a18b9725212351f19ee3aa6c7c6c/code/addons/links/src/utils.ts#L39-L62
- Template Story: https://github.com/Vazaril/storybook/blob/edb4df0e6095bd7de6303da5975f4176f0517d1f/code/addons/links/template/stories/hreftosync.stories.ts#L1-L22
- README entry: https://github.com/Vazaril/storybook/blob/edb4df0e6095bd7de6303da5975f4176f0517d1f/code/addons/links/README.md?plain=1#L97-L111

Would this added function suffice for your usecase? Feedback would be greatly appreciated :D. If this concrete implementation sounds useful to you, i would later open up a pull request for this discussed feature.

**Reply by JoshTumath** [reply]
It's very kind of you to work on this!

Yes, that looks great. My feedback is: you could extract the common code in `hrefTo` and `hrefToSync` into a new internal function so that there's less code duplication.

**Reply by Vazaril** [reply]
Thanks for your feedback :D My core idea was to change as little as possible, but the duplicated code was also a thorn in my eye. I updated my proposed solution with your feedback in mind.

**Reply by JoshTumath** [reply]
That's great! Thank you!

**Reply by Vazaril** [reply]
Just wanted to let you know that i opened a pr for this feature request thanks to your help:
https://github.com/storybookjs/storybook/pull/36189
